_source/_posts/2022-02-25-thymeleaf-security.md

```diff
@@ -0,0 +1,670 @@
+
+```java
+package com.okta.developer.demo;
+
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.web.reactive.WebFluxTest;
+import org.springframework.http.MediaType;
+import org.springframework.test.web.reactive.server.WebTestClient;
+
+import static org.springframework.security.test.web.reactive.server.SecurityMockServerConfigurers.csrf;
+import static org.springframework.security.test.web.reactive.server.SecurityMockServerConfigurers.mockOidcLogin;
+
+@WebFluxTest
+public class QuizControllerTest {
+
+    @Autowired
+    private WebTestClient client;
+
+    @Test
+    void testPostQuiz_noCSRFToken() throws Exception {
+        QuizSubmission quizSubmission = new QuizSubmission();
+        this.client.mutateWith(mockOidcLogin())
+                .post().uri("/quiz")
+                .exchange()
+                .expectStatus().isForbidden()
+                .expectBody().returnResult()
+                .toString().contains("An expected CSRF token cannot be found");
+    }
+
+    @Test
+    void testPostQuiz() throws Exception {
+        this.client.mutateWith(csrf()).mutateWith(mockOidcLogin())
+                .post().uri("/quiz")
+                .contentType(MediaType.APPLICATION_FORM_URLENCODED)
+                .exchange().expectStatus().isOk();
+    }
+
+    @Test
+    void testGetQuiz_noAuth() throws Exception {
+        this.client.get().uri("/quiz").exchange().expectStatus().is3xxRedirection();
+    }
+
+    @Test
+    void testGetQuiz() throws Exception {
+        this.client.mutateWith(mockOidcLogin())
+                .get().uri("/quiz").exchange().expectStatus().isOk();
+    }
+}
+```
+
+In the test class above, the first test `testPostQuiz_noCSRFToken()` verifies the quiz cannot be submitted without the CSRF token, even if the user did sign in. In the second test `testPostQuiz()`, the CSRF token is added to the mock request with `mutateWith(csrf())`, so the expected response status is HTTP 200 OK. The third test `testGetQuiz_noAuth` verifies a request of the quiz will be redirected (to Okta sign-in form) if the user is not authenticated. And the last test `testGetQuiz()` verifies the quiz can be accessed if the user has been authenticated with an OIDC login.
+
+Before running the application, as `quiz` is not a standard scope or a scope defined in Okta, you have to define it for the default authorization server. Sign in to the Okta dashboard, and in the left menu, go to **Security > API**, the choose the **default** authorization server. In the **Scopes** tab, click **Add Scope**. Set the scope name as `quiz` and add a description, leave all the remaining fields with default values and click on **Create**. Now the `quiz` scope can be required during the OIDC login.
+
+{% img blog/thymeleaf-security/add-scope.png alt:"Add Scope Page" width:"800" %}{: .center-image }
```

Comment:
You might want to crop this image and make it bigger. The font is kinda small.

Reply:
Cropped.